lib/at_host/at_host.c

```diff
@@ -53,49 +81,67 @@ static void at_host_at_socket_write(const void *data, size_t datalen)
 	}
 }
 
-static void uart_rx_handler(u8_t *data)
+static void at_cmd_send(u8_t *buf, size_t len)
+{
+	uart_irq_rx_disable(uart_dev);
+
+	/* Detect at_buff overflow*/
+	if (len > AT_MAX_CMD_LEN) {
+		LOG_ERR("at_buff overflow\n");
+	} else {
+		at_host_at_socket_write(buf, len);
+	}
+	uart_irq_rx_enable(uart_dev);
+}
+
+static void uart_rx_handler(struct k_work *work)
 {
 	/* Handling the characters of characters */
-	if ((*data == 127) || (*data == 8)) { /* Backspace */
+	if ((*at_cmd_read == 127) || (*at_cmd_read == 8)) { /* Backspace */
 		if (characters > 0) {
 			characters--;
 			at_buff[characters] = 0;
 		}
 	} else {
-		at_buff[characters++] = *data;
+		at_buff[characters++] = *at_cmd_read;
 	}
 	/* Detect start or end of quotes in the command. */
-	if (*data == '"') {
+	if (*at_cmd_read == '"') {
 		inside_quotes = !inside_quotes;
 	}
 	/* Ready to send the AT command. */
 	if (inside_quotes == false) {
 		switch (term_mode) {
 		case MODE_NULL_TERM:
-			if (*data == termination[term_mode]) {
-				at_rdy = true;
+			if (*at_cmd_read == termination[term_mode]) {
+				goto send;
 			}
 			break;
 		case MODE_CR:
-			if (*data == termination[term_mode]) {
-				at_rdy = true;
+			if (*at_cmd_read == termination[term_mode]) {
+				goto send;
 			}
 			break;
 		case MODE_LF:
 			if ((at_buff[characters - 2]) &&
-			    *data == termination[term_mode]) {
-				at_rdy = true;
+			    *at_cmd_read == termination[term_mode]) {
+				goto send;
 			}
 			break;
 		case MODE_CR_LF:
 			if ((at_buff[characters - 2] == '\r') &&
-			    (*data == termination[term_mode - 1])) {
-				at_rdy = true;
+			    (*at_cmd_read == termination[term_mode - 1])) {
+				goto send;
 			}
 			break;
 		default:
 			break;
 		}
+
+		return;
+send:
+		at_cmd_send(at_buff, characters);
```

Comment:
So that's where I think we should queue the work.

Reply:
Agreed, will fix